target/ppc/int_helper.c

```diff
@@ -910,6 +910,79 @@ VCT(uxs, cvtsduw, u32)
 VCT(sxs, cvtsdsw, s32)
 #undef VCT
 
+#define XVIGER_NAME(bits, rank)                                         \
+                    glue(glue(helper_xvi, bits), glue(ger, rank))
+
+#define VsrN(i) VsrB((i >> 1)) >> ((i+1) % 2)*4 & 0xF
+#define VsrSN(i) VsrB((i >> 1)) >> ((i+1) % 2)*4 & 0xF | (0xF ^ (-1))
+
+#define EXTS(num, nb)                                                   \
+             (((((num) >> (nb - 1)) & 1) == 0) ?                        \
+             num :                                                      \
+             ((((1 << nb) - 1) ^ (-1)) | (num)))
+
+#define XVIGER(bits, rank, el)                                          \
+    void XVIGER_NAME(bits, rank) (CPUPPCState *env,                     \
+                                uint32_t a_r, uint32_t b_r,             \
+                                uint32_t  at_r, uint64_t mask)          \
+    {                                                                   \
+        uint8_t pmsk, xmsk, ymsk, opmsk;                                \
+        ppc_vsr_t *a, *b, *at;                                          \
+        int64_t psum;                                                   \
+        int32_t va, vb;                                                 \
+        int i, j, k;                                                    \
+        bool sat;                                                       \
+        xmsk = (uint8_t)mask & 0x0F;                                    \
+        ymsk = (uint8_t)(mask >> 4) & 0x0F;                             \
+        pmsk = (uint8_t)(mask >> 8) & 0xFF;                             \
+        opmsk = (mask >> 32) & 0xFF;                                    \
+        a = cpu_vsr_ptr(env, a_r);                                      \
+        b = cpu_vsr_ptr(env, b_r);                                      \
+        at = cpu_vsr_ptr(env, at_r);                                    \
+        sat = (bool)(opmsk & SAT_MASK);                                 \
+        for (i = 0; i < 4; i++) {                                       \
+            for (j = 0; j < 4; j++) {                                   \
+                if ((1 << (3-i) & xmsk) && (1 << (3-j) & ymsk)) {       \
+                    if (!(opmsk & AC_MASK)) {                           \
```

Comment:
The order of these checks (acc/sat) doesn't look right.
You should first calculate psum (already including at[i].VsrW(j) when acc is set), and then chop/clamp (again, not tested :)):

```
        if ((xmsk & xmsk_bit) && (ymsk & ymsk_bit)) {
            psum = 0;
            <psum = prod0 + ... + prod<RANK-1>>
            if (acc) {
                psum += at[i].VsrW(j);
            }
            if (sat) {
                if (psum > (1LL << 31) - 1) {
                    psum = (1LL << 31) - 1;
                    set_vscr_sat(env);
                } else if (psum < -(1LL << 31)) {
                    psum = -(1LL << 31) ;
                    set_vscr_sat(env);
                }
            }
            at[i].VsrW(j) = (int32_t)psum;
```